### PR TITLE
Refactor Docker build

### LINKED ISSUE #4: Create Fedora Image
With 
- [ ] .NET 6
- [ ] .NET 5
- [ ] .NET Core 3.1

### LINKED ISSUE #5: Create Centos Image
With 
- [ ] .NET 6
- [ ] .NET 5
- [ ] .NET Core 3.1

### LINKED ISSUE #6: Create Ubuntu Image
With 
- [ ] .NET Core
- [ ] Mono

### LINKED ISSUE #7: Switch to use global tool

### LINKED ISSUE #10: Add Support for the .netCore 2.2 SDK
I have an app that I'd like to do a multi-stage docker build with cake that will be targeting netcoreapp2.2  that doesn't work as this targets the 2.1 sdk. 

The current tag `cakebuild/cake:v0.32.1-2.1-sdk` doesn't work to build a 2.2 app as that sdk isn't new enough. If we could build a set of containers targeting the `microsoft/dotnet:2.2-sdk` as the base it would be awesome.

### LINKED ISSUE #16: Add .NET Core 3 images.

### LINKED ISSUE #18: Update to latest version of Cake
Docker images should be updated to latest version of Cake. 

We should also update release documentation for Cake that updating Docker Image is part of the regular release process.

### LINKED ISSUE #22: Add Support for the .NET 6.0 SDK

### PR DESCRIPTION
**Pull request by devlead**
This PR refactors how Cake containers are built and deployed, still work in progress.

What's done
- [x] Switch to using Cake.Tool from nuget instead of building Cake again.
- [x] Switch from Azure Pipelines to GitHub Actions
- [x] Reduce number of docker files with help of using parameterized docker file
- [x] Dynamically fetched last 5 versions of Cake and build containers with that
- [x] Dynamically fetch tags from Microsoft official images, currently builds:
  - [x] Linux images
    - ~~mcr.microsoft.com/dotnet/sdk:2.1~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-alpine~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-alpine3.12~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-bionic~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-focal~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-stretch~~
    - mcr.microsoft.com/dotnet/sdk:3.1
    - mcr.microsoft.com/dotnet/sdk:3.1-alpine
    - mcr.microsoft.com/dotnet/sdk:3.1-alpine3.12
    - mcr.microsoft.com/dotnet/sdk:3.1-alpine3.13
    - mcr.microsoft.com/dotnet/sdk:3.1-alpine3.14
    - mcr.microsoft.com/dotnet/sdk:3.1-bionic
    - mcr.microsoft.com/dotnet/sdk:3.1-bullseye
    - mcr.microsoft.com/dotnet/sdk:3.1-buster
    - mcr.microsoft.com/dotnet/sdk:3.1-cbl-mariner1.0
    - mcr.microsoft.com/dotnet/sdk:3.1-focal
    - mcr.microsoft.com/dotnet/sdk:5.0
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.12
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.12-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.13
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.13-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.14
    - mcr.microsoft.com/dotnet/sdk:5.0-alpine3.14-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-bullseye-slim
    - mcr.microsoft.com/dotnet/sdk:5.0-bullseye-slim-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-buster-slim
    - mcr.microsoft.com/dotnet/sdk:5.0-buster-slim-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-cbl-mariner1.0
    - mcr.microsoft.com/dotnet/sdk:5.0-cbl-mariner1.0-amd64
    - mcr.microsoft.com/dotnet/sdk:5.0-focal
    - mcr.microsoft.com/dotnet/sdk:5.0-focal-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine3.13
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine3.13-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine3.14
    - mcr.microsoft.com/dotnet/sdk:6.0-alpine3.14-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0-bullseye-slim
    - mcr.microsoft.com/dotnet/sdk:6.0-bullseye-slim-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0-cbl-mariner1.0
    - mcr.microsoft.com/dotnet/sdk:6.0-cbl-mariner1.0-amd64
    - mcr.microsoft.com/dotnet/sdk:6.0-focal
    - mcr.microsoft.com/dotnet/sdk:6.0-focal-amd64
  - [x] Windows Images
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-nanoserver-1809~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-nanoserver-1903~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-nanoserver-1909~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-nanoserver-2004~~
    - ~~mcr.microsoft.com/dotnet/sdk:2.1-nanoserver-2009~~
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-1809
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-1903
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-1909
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-2004
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-2009
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-20H2
    - mcr.microsoft.com/dotnet/sdk:3.1-nanoserver-ltsc2022
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-1809
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-1909
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-2004
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-2009
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-20H2
    - mcr.microsoft.com/dotnet/sdk:5.0-nanoserver-ltsc2022
    - mcr.microsoft.com/dotnet/sdk:5.0-windowsservercore-ltsc2019
    - mcr.microsoft.com/dotnet/sdk:5.0-windowsservercore-ltsc2022
    - mcr.microsoft.com/dotnet/sdk:6.0-nanoserver-1809
    - mcr.microsoft.com/dotnet/sdk:6.0-nanoserver-1909
    - mcr.microsoft.com/dotnet/sdk:6.0-nanoserver-2004
    - mcr.microsoft.com/dotnet/sdk:6.0-nanoserver-20H2
    - mcr.microsoft.com/dotnet/sdk:6.0-nanoserver-ltsc2022
    - mcr.microsoft.com/dotnet/sdk:6.0-windowsservercore-ltsc2019
    - mcr.microsoft.com/dotnet/sdk:6.0-windowsservercore-ltsc2022
  - [x] Currently exluded tags  
    - `5.0-alpine3.11` at testing lacked RTM 5 SDK
    - `5.0-nanoserver-1903` at testing lacked RTM 5 SDK
- [x] Create Ubuntu Image 
- [x] Enable docker publish on main

What remains to be done
- [x] Tag "latest" version
- [ ] Create Centos Image 
- [ ] Create Fedora Image 
- [ ] ~~Mono images based on above built~~
- [ ] Bitrise images based on above built
- [x] Add image filter as commandline argument
  - [x] Build any images that can on `windows-latest`
- [x] Update readme
- [ ] Upate docs - _separate PR to website_
- [ ] Upate blog - _separate PR to website_

Related issues
- [x] fixes #18 Update to latest version of Cake
- [x] fixes #16 Add .NET Core 3 images.
- [x] fixes #7 Switch to use global tool
- [x] fixes #6 Create Ubuntu Image
- [x] closes #10 ~~Add Support for the .netCore 2.2 SDK~~ - _Won't fix as it's no longer supported by Microsoft_
- [x] fixes #22

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,50 +1,124 @@
...
-|cakebuild/cake:2.1-sdk-mono            | Image with latest released Cake and Mono,  based on cakebuild/cake:2.1-sdk                        | [![2.1-sdk-mono Size](https://img.shields.io/microbadger/image-size/cakebuild/cake/2.1-sdk-mono.svg)](https://hub.docker.com/r/cakebuild/cake/tags/)                  |
-|cakebuild/cake:2.1-sdk-bitrise-mono    | Image with latest released Cake, Bitrise CLI and Mono,  based on cakebuild/cake:2.1-sdk-bitrise   | [![2.1-sdk-bitrise-mono size](https://img.shields.io/microbadger/image-size/cakebuild/cake/2.1-sdk-bitrise-mono.svg)](https://hub.docker.com/r/cakebuild/cake/tags/)  |
+Cake official Docker files with .NET SDK and Cake global tool pre-installed.
+
+## Images
+
+Images are currently continuously built for last 10 versions of Cake version 1.0 or newer.
+Currently based on official Microsoft images available ( [mcr.microsoft.com/dotnet/sdk](https://github.com/microsoft/containerregistry) ).
+
+To to pin to a specific version suffix with Cake version i.e. `cakebuild/cake:sdk-6.0` becomes `cakebuild/cake:sdk-6.0-v1.3.0`.
```
Comment (gep13) [inline README.md:12]:
```suggestion
To pin to a specific version suffix with Cake version i.e. `cakebuild/cake:sdk-6.0` becomes `cakebuild/cake:sdk-6.0-v1.3.0`.
```

File: README.md [hunk 2]
```diff
@@ -1,50 +1,124 @@
...
+|-----------------------|---------------------------|-------------------------------|
+| CAKE_PATHS_TOOLS      | /cake_build/tools         | C:/cake_build/tools           |
+| CAKE_PATHS_ADDINS     | /cake_build/tools/Addins  | C:/cake_build/tools/Addins    |
+| CAKE_PATHS_MODULES    | /cake_build/tools/Modules | C:/cake_build/tools/Modules   |
+
+These are set to be able to keep foreign binaries inside your container and also enable caching of tools, addins and modules between docker layers.
+
+## Build Infrastructure
+
+Containers are currently build using GitHub Actions.
```
Comment (gep13) [inline README.md:108]:
```suggestion
Containers are currently built using GitHub Actions.
```

File: README.md [hunk 3]
```diff
@@ -1,50 +1,124 @@
...
+But you can also pass parameters to filter to specific base images to build or exclude individual to not build.
 
-I.e. if `v0.30.0` is latest release Cake version then it'll be tagged `2.1-sdk` and `v0.30.0-2.1-sdk`, when `v0.31.0` is released then it'll be tagged `2.1-sdk` and `v0.31.0-2.1-sdk`, and `v0.30.0-2.1-sdk` will be untouched.
+### Build script parameters
 
-Currently the images come pre-installed with a self-contained version of Cake.CoreCLR, this means modules, addins and tools need to be .NET Core or .NET Standard 2.0 compatible.
\ No newline at end of file
+| Name                      | Description                                                                                           |
+|---------------------------|-------------------------------------------------------------------------------------------------------|
+| remove-base-image         | Remove mcr base image after built, useful to conserve space used when building                        |
```
Comment (gep13) [inline README.md:122]:
```suggestion
| remove-base-image         | Remove mcr base image after build, useful to conserve space used when building                        |
```

File: build/helpers/docker.cake [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+public static Policy DockerRetryPolicy;
+DockerRetryPolicy = Policy
+                    .Handle<Exception>()
+                    .WaitAndRetry(5,
+                        retryAttempt => TimeSpan.FromSeconds(Math.Pow(2, 2 + retryAttempt)),
+                        (exception, timeSpan, retryCount, context)=>{
+                            Context.Log.Warning("Retry {0} because {1}...", retryCount, exception.Message);
+                        });
+
+FilePath        dockerPath          =  Context.Tools.Resolve(IsRunningOnWindows() ? "docker.exe" : "docker")
```
Comment (gep13) [inline build/helpers/docker.cake:17]: The spacing seems a little off here...

**Review by gep13 — CHANGES_REQUESTED** [review]
A couple minor changes, otherwise LGTM!

File: build/helpers/docker.cake (hunk 4 above)
Comment (devlead) [reply build/helpers/docker.cake]: adjusted.

**Pull request merged**